Build failure: ponyc on aarch64-linux

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

Try to build `ponyc` for `aarch64-linux`.

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/295981449

### Relevant log output

```console
<snip>
[100%] Linking CXX executable /build/source/build/release/libponyc.tests
warning: Skipping impure flag -march=native because NIX_ENFORCE_NO_NATIVE is set
CMakeFiles/libponyc.tests.dir/compiler_serialisation.cc.o: in function `CompilerSerialisationTest::test_pass_ast(char const*)':
compiler_serialisation.cc:(.text+0x48): relocation truncated to fit: R_AARCH64_LD64_GOTPAGE_LO15 against symbol `__stack_chk_guard@@GLIBC_2.17' defined in .data.rel.ro section in /nix/store/z3nhrl96ba4ccrvl6r4c259g58giiicl-glibc-2.40-66/lib/ld-linux-aarch64.so.1
/nix/store/ix9mh0yqi95fhnqw87sww35jgjwl7wg1-binutils-2.44/bin/ld: compiler_serialisation.cc:(.text+0x48): warning: too many GOT entries for -fpic, please recompile with -fPIC
/nix/store/ix9mh0yqi95fhnqw87sww35jgjwl7wg1-binutils-2.44/bin/ld: final link failed
collect2: error: ld returned 1 exit status
make[3]: *** [test/libponyc/CMakeFiles/libponyc.tests.dir/build.make:968: /build/source/build/release/libponyc.tests] Error 1
make[2]: *** [CMakeFiles/Makefile2:588: test/libponyc/CMakeFiles/libponyc.tests.dir/all] Error 2
make[1]: *** [Makefile:136: all] Error 2
make[1]: Leaving directory '/build/source/build/build_release'
make: *** [Makefile:195: build] Error 2
```

### Additional context

_No response_

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.20, NixOS, 25.05 (Warbler), 25.05.20250428.a80e650`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.28.2`
 - nixpkgs: `/nix/store/yhkbsp6jl23gk35bjmb28yn8q0ghb9dk-source`

### Notify maintainers

@kamilchm 
@patternspandemic
@redvers
@numinit
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/po/ponyc/disable-networking-tests.patch
+++ b/pkgs/by-name/po/ponyc/disable-networking-tests.patch
@@ -1,16 +1,42 @@
+From e49f97eb4b0fd27b26437638db7984fbcfd3a14f Mon Sep 17 00:00:00 2001
+From: Morgan Jones <me@numin.it>
+Date: Sun, 4 May 2025 15:46:07 -0700
+Subject: [PATCH] net: disable tests in the Nix sandbox
+
+---
+ packages/net/_test.pony | 19 +------------------
+ 1 file changed, 1 insertion(+), 18 deletions(-)
+
 diff --git a/packages/net/_test.pony b/packages/net/_test.pony
-index 9044dfb1..f0ea10f7 100644
+index 05462eb2..c0c6cdfa 100644
 --- a/packages/net/_test.pony
 +++ b/packages/net/_test.pony
-@@ -26,11 +26,6 @@ actor \nodoc\ Main is TestList
-       test(_TestTCPThrottle)
-     end
+@@ -15,25 +15,8 @@ actor \nodoc\ Main is TestList
+   new make() => None
  
+   fun tag tests(test: PonyTest) =>
+-    // Tests below function across all systems and are listed alphabetically
++    // (@numinit): only this test works in the Nix sandbox:
+     test(_TestTCPConnectionFailed)
+-    test(_TestTCPExpect)
+-    test(_TestTCPExpectOverBufferSize)
+-    test(_TestTCPMute)
+-    test(_TestTCPProxy)
+-    test(_TestTCPUnmute)
+-    test(_TestTCPWritev)
+-
+-    // Tests below exclude windows and are listed alphabetically
+-    ifdef not windows then
+-      test(_TestTCPConnectionToClosedServerFailed)
+-      test(_TestTCPThrottle)
+-    end
+-
 -    // Tests below exclude osx and are listed alphabetically
 -    ifdef not osx then
 -      test(_TestBroadcast)
 -    end
--
+ 
  class \nodoc\ _TestPing is UDPNotify
    let _h: TestHelper
-   let _ip: NetAddress
+-- 
+2.47.0

--- a/pkgs/by-name/po/ponyc/disable-process-tests.patch
+++ b/pkgs/by-name/po/ponyc/disable-process-tests.patch
@@ -1,0 +1,26 @@
+From 77d703b11d298f6be88b04f7e8ca85de139e82be Mon Sep 17 00:00:00 2001
+From: Morgan Jones <me@numin.it>
+Date: Mon, 5 May 2025 20:34:02 -0700
+Subject: [PATCH] process: disable KillLongRunningChild test
+
+---
+ packages/process/_test.pony | 3 ++-
+ 1 file changed, 2 insertions(+), 1 deletion(-)
+
+diff --git a/packages/process/_test.pony b/packages/process/_test.pony
+index fe9fdb04..756588f9 100644
+--- a/packages/process/_test.pony
++++ b/packages/process/_test.pony
+@@ -18,7 +18,8 @@ actor \nodoc\ Main is TestList
+     test(_TestChdir)
+     test(_TestExpect)
+     test(_TestFileExecCapabilityIsRequired)
+-    test(_TestKillLongRunningChild)
++    // (@booxter/@numinit) Appears to be flaky.
++    // test(_TestKillLongRunningChild)
+     test(_TestLongRunningChild)
+     test(_TestNonExecutablePathResultsInExecveError)
+     test(_TestPrintvOrdering)
+-- 
+2.47.0
+

--- a/pkgs/by-name/po/ponyc/package.nix
+++ b/pkgs/by-name/po/ponyc/package.nix
@@ -19,6 +19,7 @@
   which,
   z3,
   cctools,
+  procps,
 }:
 
 stdenv.mkDerivation (rec {
@@ -73,6 +74,7 @@ stdenv.mkDerivation (rec {
     [
       # Sandbox disallows network access, so disabling problematic networking tests
       ./disable-networking-tests.patch
+      ./disable-process-tests.patch
     ]
     ++ lib.optionals stdenv.hostPlatform.isDarwin [
       (replaceVars ./fix-darwin-build.patch {
@@ -102,10 +104,19 @@ stdenv.mkDerivation (rec {
         --replace-fail "https://github.com/google/googletest/archive/refs/tags/v$googletestRev.tar.gz" "$NIX_BUILD_TOP/deps/googletest-$googletestRev.tar"
   '';
 
-  preBuild = ''
-    make libs build_flags=-j$NIX_BUILD_CORES
-    make configure build_flags=-j$NIX_BUILD_CORES
-  '';
+  preBuild =
+    ''
+      extraFlags=(build_flags=-j$NIX_BUILD_CORES)
+    ''
+    + lib.optionalString stdenv.hostPlatform.isAarch64 ''
+      # See this relnote about building on Raspbian:
+      # https://github.com/ponylang/ponyc/blob/0.46.0/.release-notes/0.45.2.md
+      extraFlags+=(pic_flag=-fPIC)
+    ''
+    + ''
+      make libs "''${extraFlags[@]}"
+      make configure "''${extraFlags[@]}"
+    '';
 
   makeFlags = [
     "PONYC_VERSION=${version}"
@@ -117,8 +128,9 @@ stdenv.mkDerivation (rec {
     "-Wno-error=implicit-fallthrough"
   ];
 
-  # make: *** [Makefile:222: test-full-programs-release] Killed: 9
-  doCheck = !stdenv.hostPlatform.isDarwin;
+  doCheck = true;
+
+  nativeCheckInputs = [ procps ];
 
   installPhase =
     ''

